File: CulinarySite/ServiceLayer/RecipeOrganicMatterService.cs
using Database;
using Repositories;
using System.Collections.Generic;
using ServiceLayer.ViewModels.RecipeOrganicMatter;

namespace ServiceLayer
{
    public class RecipeOrganicMatterService : IRecipeOrganicMatterService
    {
        private readonly IReadOnlyGenericRepository<RecipeOrganicMatter> recipeOrganicMatterReadOnlyRepository;
        private readonly IWriteGenericRepository<RecipeOrganicMatter> recipeOrganicMatterWriteRepository;
        public RecipeOrganicMatterService(
            IReadOnlyGenericRepository<RecipeOrganicMatter> recipeOrganicMatterReadOnlyRepository,
            IWriteGenericRepository<RecipeOrganicMatter> recipeOrganicMatterWriteRepository)
        {
            this.recipeOrganicMatterReadOnlyRepository = recipeOrganicMatterReadOnlyRepository;
            this.recipeOrganicMatterWriteRepository = recipeOrganicMatterWriteRepository;
        }
        public void CreateRecipeOrganicMatter(CreateRecipeOrganicMatterModel createRecipeOrganicMatterModel)
        {
            var recipeOrganicMatter = new RecipeOrganicMatter
            {
                OrganicMatterId = createRecipeOrganicMatterModel.OrganicMatterId,
                Unit = createRecipeOrganicMatterModel.Unit,
                Quantity = createRecipeOrganicMatterModel.Quantity,
            };
            this.recipeOrganicMatterWriteRepository.Create(recipeOrganicMatter);
            this.recipeOrganicMatterWriteRepository.Save();
        }

        public void UpdateRecipeOrganicMatter(UpdateRecipeOrganicMatterModel updateRecipeOrganicMatterModel)
        {
            var recipeOrganicMatter = new RecipeOrganicMatter
            {
                Id = updateRecipeOrganicMatterModel.RecipeOrganicMatterId,
                OrganicMatterId = updateRecipeOrganicMatterModel.OrganicMatterId,
                Unit = updateRecipeOrganicMatterModel.Unit,
                Quantity = updateRecipeOrganicMatterModel.Quantity,
            };
            this.recipeOrganicMatterWriteRepository.Update(recipeOrganicMatter);
            this.recipeOrganicMatterWriteRepository.Save();
        }

        public void DeleteRecipeOrganicMatter(int id)
        {
            this.recipeOrganicMatterWriteRepository.Delete(id);
            this.recipeOrganicMatterWriteRepository.Save();
        }

        public IEnumerable<RecipeOrganicMatterListModel> GetRecipeOrganicMatterList(bool withRelated)
        {
            IEnumerable<RecipeOrganicMatter> recipeOrganicMatters;
            List<RecipeOrganicMatterListModel> recipeOrganicMatterListModels = new List<RecipeOrganicMatterListModel>();
            if (withRelated)
            {
                recipeOrganicMatters = this.recipeOrganicMatterReadOnlyRepository.GetItemListWithInclude(
                    x => x.OrganicMatter,
                    x => x.Recipes);
                foreach (var recipeOrganicMatter in recipeOrganicMatters)
                {
                    recipeOrganicMatterListModels.Add(new RecipeOrganicMatterListModel
                    {
                        RecipeOrganicMatterId = recipeOrganicMatter.Id,
                        OrganicMatterName = recipeOrganicMatter.OrganicMatter.Name,
                        Unit = recipeOrganicMatter.Unit,
                        Quantity = recipeOrganicMatter.Quantity,
                    });
                }
                return recipeOrganicMatterListModels;
            }
            recipeOrganicMatters = this.recipeOrganicMatterReadOnlyRepository.GetItemList();
            foreach (var recipeOrganicMatter in recipeOrganicMatters)
            {
                recipeOrganicMatterListModels.Add(new RecipeOrganicMatterListModel
                {
                    RecipeOrganicMatterId = recipeOrganicMatter.Id,
                    Unit = recipeOrganicMatter.Unit,
                    Quantity = recipeOrganicMatter.Quantity
                });
            }
            return recipeOrganicMatterListModels;
        }

        public RecipeOrganicMatterDetailModel GetRecipeOrganicMatter(int id, bool withRelated)
        {
            RecipeOrganicMatter recipeOrganicMatter = new RecipeOrganicMatter();
            RecipeOrganicMatterDetailModel recipeOrganicMatterDetailModel = new RecipeOrganicMatterDetailModel();

            if (withRelated)
            {
                recipeOrganicMatter = this.recipeOrganicMatterReadOnlyRepository.GetItemWithInclude(
                    x => x.Id == id,
                    x => x.OrganicMatter,
                    x => x.Recipes);
                recipeOrganicMatterDetailModel = new RecipeOrganicMatterDetailModel
                {
                    RecipeOrganicMatterId = recipeOrganicMatter.Id,
                    OrganicMatterId = recipeOrganicMatter.OrganicMatterId,
                    Unit = recipeOrganicMatter.Unit,
                    Quantity = recipeOrganicMatter.Quantity,
                };
                return recipeOrganicMatterDetailModel;
            }
            recipeOrganicMatter = this.recipeOrganicMatterReadOnlyRepository.GetItem(id);
            recipeOrganicMatterDetailModel = new RecipeOrganicMatterDetailModel
            {
                RecipeOrganicMatterId = recipeOrganicMatter.Id,
                Unit = recipeOrganicMatter.Unit,
                Quantity = recipeOrganicMatter.Quantity
            };
            return recipeOrganicMatterDetailModel;
        }
    }
}

File: CulinarySiteApp/src/app/services/recipe-organic-matter.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { IOrganicMatterListModel } from '../interfaces/organic-matter/organic-matter-list-model.interface';
import { IOrganicMatterDetailModel } from '../interfaces/organic-matter/organic-matter-detail-model.interface';
import { CreateOrganicMatterModel } from '../viewmodels/organic-matter/create-organic-matter-model.class';
import { UpdateOrganicMatterModel } from '../viewmodels/organic-matter/update-organic-matter-model.class';

@Injectable()
export class RecipeOrganicMatterService {
  public url: string = '/api/recipeorganicmatter';

  constructor(private http: HttpClient) {}

  public getRecipeOrganicMatterList(
    withRelated: boolean
  ): Observable<IOrganicMatterListModel[]> {
    return this.http.get<IOrganicMatterListModel[]>(
      this.url + '/' + withRelated
    );
  }

  public getRecipeOrganicMatter(
    id: number,
    withRelated: boolean
  ): Observable<IOrganicMatterDetailModel> {
    return this.http.get<IOrganicMatterDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createRecipeOrganicMatter(
    createRecipeOrganicMatterModel: CreateOrganicMatterModel
  ): Observable<void> {
    return this.http.post<void>(this.url, createRecipeOrganicMatterModel);
  }

  public updateRecipeOrganicMatter(
    updateRecipeOrganicMatterModel: UpdateOrganicMatterModel
  ): Observable<void> {
    return this.http.put<void>(this.url, updateRecipeOrganicMatterModel);
  }

  public deleteRecipeOrganicMatter(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}
